internal/cachecontroller/cache_controller_test.go

```diff
@@ -85,149 +88,209 @@ func TestInMemoryCacheController_findChangesAndInvalidate(t *testing.T) {
+							TupleKey: &openfgav1.TupleKey{
+								Object:   "test:7",
+								Relation: "writer",
+								User:     "test",
+							},
+						},
+						{
+							Operation: openfgav1.TupleOperation_TUPLE_OPERATION_WRITE,
+							Timestamp: timestamppb.New(time.Now().Add(-12 * time.Second)),
+							TupleKey: &openfgav1.TupleKey{
+								Object:   "test:8",
+								Relation: "writer",
+								User:     "test",
+							},
+						},
+					}, "", nil),
+					cache.EXPECT().Get(storage.GetChangelogCacheKey("6")).Return(&storage.ChangelogCacheEntry{LastModified: time.Now().Add(-20 * time.Second)}),
+					cache.EXPECT().Set(storage.GetChangelogCacheKey("6"), gomock.Any(), gomock.Any()),
+					cache.EXPECT().Set(storage.GetInvalidIteratorByObjectRelationCacheKeys("6", "test:5", "viewer")[0], gomock.Any(), gomock.Any()),
+					cache.EXPECT().Set(storage.GetInvalidIteratorByUserObjectTypeCacheKeys("6", []string{"test"}, "test")[0], gomock.Any(), gomock.Any()),
+				)
+			},
 		},
 		{
-			name:               "last_change_not_in_newest_batch",
-			storeID:            "7",
-			readChangesResults: &readChangesResponse{err: nil, changes: generateChanges("test", "relation", "user", 50)},
-			setCacheKeys: []string{
-				storage.GetChangelogCacheKey("7"),
-				storage.GetInvalidIteratorCacheKey("7")},
+			name:    "last_change_not_in_newest_batch",
+			storeID: "7",
+			setMocks: func(cache *mocks.MockInMemoryCache[any], datastore *mocks.MockOpenFGADatastore) {
+				gomock.InOrder(
+					datastore.EXPECT().ReadChanges(gomock.Any(), "7", gomock.Any(), gomock.Any()).Return(
+						generateChanges("test", "relation", "user", 50), "", nil),
+					cache.EXPECT().Get(storage.GetChangelogCacheKey("7")).Return(&storage.ChangelogCacheEntry{LastModified: time.Now().Add(-20 * time.Second)}),
+					cache.EXPECT().Set(storage.GetChangelogCacheKey("7"), gomock.Any(), gomock.Any()),
+					cache.EXPECT().Set(storage.GetInvalidIteratorCacheKey("7"), gomock.Any(), gomock.Any()),
+				)
+			},
+		},
+		{
+			name:    "initial_check_for_invalidation",
+			storeID: "8",
+			setMocks: func(cache *mocks.MockInMemoryCache[any], datastore *mocks.MockOpenFGADatastore) {
+				gomock.InOrder(
+					datastore.EXPECT().ReadChanges(gomock.Any(), "8", gomock.Any(), gomock.Any()).Return([]*openfgav1.TupleChange{
+						{
+							Operation: openfgav1.TupleOperation_TUPLE_OPERATION_WRITE,
+							Timestamp: timestamppb.New(time.Now().Add(-20 * time.Second)),
+							TupleKey: &openfgav1.TupleKey{
+								Object:   "test:8",
+								Relation: "viewer",
+								User:     "test",
+							}},
+					}, "", nil),
+					cache.EXPECT().Get(storage.GetChangelogCacheKey("8")).Return(nil),
+					cache.EXPECT().Set(storage.GetChangelogCacheKey("8"), gomock.Any(), gomock.Any()),
+					cache.EXPECT().Set(storage.GetInvalidIteratorCacheKey("8"), gomock.Any(), gomock.Any()),
+				)
+			},
+		},
+		{
+			name:    "initial_check_for_invalidation_change_is_recent",
```

Comment:
The difference between this testcase and `initial_check_for_invalidation`, is only the Timestamp of the tuples. Could you clarify how this matters?

Reply:
added note in code comment but this is more as response to your feedback in https://github.com/openfga/openfga/pull/2190#discussion_r1905963147 regarding you cannot find the double negative case in test.